tool/net/redbean.c

```diff
@@ -2579,6 +2579,22 @@ static void TuneServerSocket(void) {
   LOGIFNEG1(setsockopt(server, IPPROTO_TCP, TCP_QUICKACK, &yes, sizeof(yes)));
 }
 
+static void OpenBrowser(const char *serveraddrname) {
+  char openbrowsercommand[255];
+  char *prog;
+  if (IsWindows()) {
+    prog = "explorer";
+  } else if (IsXnu()) {
+    prog = "open";
+  } else {
+    prog = "xdg-open";
+  }
+  snprintf(openbrowsercommand, sizeof(openbrowsercommand), "%s http://%s", prog,
```

Comment:
One issue with `serveraddrname` is it might be `0.0.0.0:8080`. What I'd recommend doing is always use 127.0.0.1 and then the port can be obtained via `ntohs(serveraddr.sin_port)`.

Reply:
Also a note on 0.0.0.0 is that yeah that is serveraddrstr resolves to, but the server still works with 0.0.0.0 as well as 127.0.0.1. After doing some reading - 0.0.0.0 is "every local ip on this machine" whereas 127.0.0.1 is just "loopback" or localhost. Should I still change it to 127.0.0.1 for functional reasons? or is it just a cosmetic thing.

Reply:
Oh wow you're correct. I knew `0.0.0.0` would `bind()` to all local interfaces. I never considered it would work on the client side with `connect()`. Linux and FreeBSD appear to explicitly define `connect(0.0.0.0)` as `connect(127.0.0.1)` but that does not appear to be the case in all the os / browser / client combinations I've tried. I also just realized the user might explicitly bind to `192.168.0.x` or something. So the right thing to do here might be:

```c
  uint32_t addr;
  char ip4buf[16];
  addr = serveraddr.sin_addr.s_addr;
  if (ntohl(addr) == INADDR_ANY) addr = htonl(INADDR_LOOPBACK);
  snprintf(openbrowsercommand, sizeof(openbrowsercommand), "%s http://%s:%d\n",
           prog, inet_ntop(AF_INET, &addr, ip4buf, sizeof(ip4buf)),
           ntohs(serveraddr.sin_port));
```

Reply:
Alright, I've added this new code, removed the -b flags and tested multiple different `-l` flags (127.0.0.1 and 127.0.0.2) on Ubuntu 20.04, browser launches at the right addresses.
I guess this needs testing on windows and BSD etc?